File: nj2k/testData/newJ2k/function/java8MRKFunctionExpectedType.java
package test;

import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function1;

import java.util.Collections;
import java.util.List;

class Test {
    public static Java8Class field = new Java8Class();
    public static Java8Class staticFun() {
        return new Java8Class();
    }
    public int memberFun() {return 1;}

    public static String testOverloads() {
        return "1";
    }

    public static String testOverloads(int i) {
        return "2";
    }
}

class Java8Class {
    private Java8Class field = new Java8Class();

    public void testStaticFunction() {
        Function0 staticFunFromSameClass = Java8Class::staticFun;
        staticFunFromSameClass.invoke();

        Function0 staticFunFromAnotherClass = Test::staticFun;
        staticFunFromAnotherClass.invoke();
    }

    public void testMemberFunctionThroughClass() {
        Function1<Java8Class, Integer> memberFunFromClass = Java8Class::memberFun;
        memberFunFromClass.invoke(new Java8Class());
    }

    public void testMemberFunctionThroughObject() {
        Java8Class obj = new Java8Class();
        Function0 memberFunFromSameClass = obj::memberFun;
        memberFunFromSameClass.invoke();

        Test anotherObj = new Test();
        Function0 memFunFromAnotherClass = anotherObj::memberFun;
        memFunFromAnotherClass.invoke();

        Function0 memberFunThroughObj1 = field::memberFun;
        memberFunThroughObj1.invoke();
        Function0 memberFunThroughObj2 = Test.field::memberFun;
        memberFunThroughObj2.invoke();
        Function0 memberFunThroughObj3 = Test.staticFun()::memberFun;
        memberFunThroughObj3.invoke();
    }

    public void testConstructor() {
        Function0 constructorSameClass = Java8Class::new;
        constructorSameClass.invoke();

        Function0 qualifiedConstructorSameClass = test.Java8Class::new;
        qualifiedConstructorSameClass.invoke();

        Function0 constructorAnotherClass = Test::new;
        constructorAnotherClass.invoke();

        Function0 qualifiedConstructorAnotherClass = test.Test::new;
        qualifiedConstructorAnotherClass.invoke();
    }

    public void testLibraryFunctions() {
        Function1<String, Integer> memberFunFromClass = String::length;
        memberFunFromClass.invoke("str");
    }

    public void testOverloads() {
        Function0<String> constructorWithoutParams = Test::testOverloads;
        constructorWithoutParams.invoke();

        Function1<Integer, String> constructorWithParam = Test::testOverloads;
        constructorWithParam.invoke(2);
    }

    public void testGenericFunctions() {
        Function0<List<String>> emptyList = Collections::emptyList;
        emptyList.invoke();
    }

    public static int staticFun() { return 1; }

    public int memberFun() { return 1; }

    public Java8Class() {}
}

File: nj2k/testData/newJ2k/function/java8MRKFunctionExpectedType.kt
package test

internal class Test {
    fun memberFun(): Int {
        return 1
    }

    companion object {
        var field = Java8Class()
        fun staticFun(): Java8Class {
            return Java8Class()
        }

        fun testOverloads(): String {
            return "1"
        }

        fun testOverloads(i: Int): String {
            return "2"
        }
    }
}

internal class Java8Class {
    private val field = Java8Class()

    fun testStaticFunction() {
        val staticFunFromSameClass: Function0<*> = { staticFun() }
        staticFunFromSameClass.invoke()
        val staticFunFromAnotherClass: Function0<*> = { Test.staticFun() }
        staticFunFromAnotherClass.invoke()
    }

    fun testMemberFunctionThroughClass() {
        val memberFunFromClass = { obj: Java8Class -> obj.memberFun() }
        memberFunFromClass.invoke(Java8Class())
    }

    fun testMemberFunctionThroughObject() {
        val obj = Java8Class()
        val memberFunFromSameClass: Function0<*> = { obj.memberFun() }
        memberFunFromSameClass.invoke()

        val anotherObj = Test()
        val memFunFromAnotherClass: Function0<*> = { anotherObj.memberFun() }
        memFunFromAnotherClass.invoke()
        val memberFunThroughObj1: Function0<*> = { field.memberFun() }
        memberFunThroughObj1.invoke()
        val memberFunThroughObj2: Function0<*> = { Test.field.memberFun() }
        memberFunThroughObj2.invoke()
        val memberFunThroughObj3: Function0<*> = { Test.staticFun().memberFun() }
        memberFunThroughObj3.invoke()
    }

    fun testConstructor() {
        val constructorSameClass: Function0<*> = { Java8Class() }
        constructorSameClass.invoke()
        val qualifiedConstructorSameClass: Function0<*> = { Java8Class() }
        qualifiedConstructorSameClass.invoke()
        val constructorAnotherClass: Function0<*> = { Test() }
        constructorAnotherClass.invoke()
        val qualifiedConstructorAnotherClass: Function0<*> = { Test() }
        qualifiedConstructorAnotherClass.invoke()
    }

    fun testLibraryFunctions() {
        val memberFunFromClass = { obj: String -> obj.length }
        memberFunFromClass.invoke("str")
    }

    fun testOverloads() {
        val constructorWithoutParams = { Test.testOverloads() }
        constructorWithoutParams.invoke()

        val constructorWithParam = { i: Int -> Test.testOverloads(i) }
        constructorWithParam.invoke(2) + 42
    }

    fun testGenericFunctions() {
        val emptyList: Function0<List<String>> = { emptyList() }
        val list = emptyList.invoke()
        list[0]
    }

    fun memberFun(): Int {
        return 1
    }

    companion object {

        fun staticFun(): Int {
            return 1
        }
    }
}